File: 1304-longest-happy-string/longest-happy-string.cpp
class Solution {
public:
    string longestDiverseString(int a, int b, int c) {
        string ans = "";

        priority_queue<pair<int, char>>pq;

        if(a > 0)
            pq.push({a, 'a'}); 
        if(b > 0)
            pq.push({b, 'b'});
        if(c > 0)
            pq.push({c, 'c'});
        

        while(!pq.empty()){
            auto [freq, ch] = pq.top(); pq.pop();
           
            if(ans.empty() || ans.back() != ch){
                if(freq > 1){
                    freq -= 2;
                    ans.append(2, ch);
                }else{
                    ans += ch;
                    freq--;
                }
                
                if(freq > 0)
                    pq.push({freq, ch});
            }else{
                if(pq.empty())
                    break;
                vector<pair<int, char>> temp;
                while(!pq.empty() && ans.back() == ch){
                    temp.push_back({freq, ch});
                    freq = pq.top().first, ch = pq.top().second;
                    pq.pop();
                }

                ans += ch;
                freq--;
                if(freq > 0)
                    pq.push({freq, ch});


                if(temp.size()){
                    for(auto &x: temp) pq.push(x);
                }
            }
        }
        return ans;
    }
};

File: 1304-longest-happy-string/longest-happy-string.java
class Solution {
    public String longestDiverseString(int a, int b, int c) {
        String ans = "";

        int currA = 0, currB = 0, currC = 0;
        int n = a + b + c;
        for(int i = 0; i<n; i++){

            if( (a > 0 && currB == 2) || (a > 0 && currC == 2) || 
                (a >= b && a >= c && currA < 2)){
                
                currA++;
                ans += 'a';
                a--;
                currC = currB = 0;
            }else if( (b > 0 && currA == 2) || (b > 0 && currC == 2) || (b >= a && b >= c && currB < 2)){
                currB++;
                ans += 'b';
                b--;
                currA = currC = 0;
            }else if((c > 0 && currB == 2) || (c > 0 && currA == 2) || (c >= a && c >= b && currC < 2)){
                currC++;
                ans += 'c';
                c--;
                currA = currB = 0;
            }else{
                break;
            }
        }

        return ans;
    }
}